ddtrace/contrib/sanic/patch.py

```diff
@@ -136,42 +163,97 @@ def unwrap(request, write_callback=None, stream_callback=None, **kwargs):
+
+        return await wrapped(request, **new_kwargs)
+
+
+def _create_sanic_request_span(request):
+    """Helper to create sanic.request span and attach a pin to request.ctx"""
     pin = Pin()
-    if SANIC_PRE_21:
+    pin.onto(request.ctx)
+
+    if SANIC_VERSION < (21, 0, 0):
         # Set span resource from the framework request
         resource = "{} {}".format(request.method, _get_path(request))
     else:
         # The path is not available anymore in 21.x. Get it from
         # the _run_request_middleware instrumented method.
         resource = None
-        pin.onto(request.ctx)
 
     headers = request.headers.copy()
 
     trace_utils.activate_distributed_headers(ddtrace.tracer, int_config=config.sanic, request_headers=headers)
 
-    with pin.tracer.trace(
+    span = pin.tracer.trace(
         "sanic.request",
         service=trace_utils.int_service(None, config.sanic),
         resource=resource,
         span_type=SpanTypes.WEB,
-    ) as span:
-        sample_rate = config.sanic.get_analytics_sample_rate(use_global_config=True)
-        if sample_rate is not None:
-            span.set_tag(ANALYTICS_SAMPLE_RATE_KEY, sample_rate)
-
-        method = request.method
-        url = "{scheme}://{host}{path}".format(scheme=request.scheme, host=request.host, path=request.path)
-        query_string = request.query_string
-        if isinstance(query_string, bytes):
-            query_string = query_string.decode()
-        trace_utils.set_http_meta(
-            span, config.sanic, method=method, url=url, query=query_string, request_headers=headers
-        )
+    )
+    sample_rate = config.sanic.get_analytics_sample_rate(use_global_config=True)
+    if sample_rate is not None:
+        span.set_tag(ANALYTICS_SAMPLE_RATE_KEY, sample_rate)
 
-        if write_callback is not None:
-            new_kwargs["write_callback"] = _wrap_response_callback(span, write_callback)
-        if stream_callback is not None:
-            new_kwargs["stream_callback"] = _wrap_response_callback(span, stream_callback)
+    method = request.method
+    url = "{scheme}://{host}{path}".format(scheme=request.scheme, host=request.host, path=request.path)
+    query_string = request.query_string
+    if isinstance(query_string, bytes):
+        query_string = query_string.decode()
+    trace_utils.set_http_meta(span, config.sanic, method=method, url=url, query=query_string, request_headers=headers)
 
-        return await wrapped(request, **new_kwargs)
+    return span
+
+
+async def sanic_http_lifecycle_handle(request):
+    """Lifecycle signal called when a new request is started."""
+    _create_sanic_request_span(request)
+
+
+async def sanic_http_routing_after(request, route, kwargs, handler):
+    """Lifecycle signal called after routing has been resolved."""
+    span = _get_current_span(request)
+    if not span:
+        return
+
+    pattern = route.raw_path
+    if not pattern.startswith("/"):
+        pattern = "/{}".format(pattern)
```

Comment:
Why do we do this? Could it be misleading since we might report a resource that doesn't match the actual pattern used?

Reply:
sanic strips the leading slash from the path here, so we end up with `GET hello` instead of `GET /hello`.

I did this to try and make it be as close to a URL path as possible / consistent with other web frameworks.

I do not feel strongly about this, so happy to remove this check.

Reply:
The concern would be if sanic exposes this same data somehow and then it doesn't line up with logs/exception messages etc which might cause confusion when triaging.

Reply:
I added a comment, Sanic < 21.9.0 has the slash, and >= 21.9.0 strip it.

Adding the slash back here is for consistency with older Sanic versions.